examples/vectoradd_torch/tesseract_api.py

```diff
@@ -0,0 +1,191 @@
+    return out
+
+
+#
+# Pytorch-handled AD endpoints (no need to modify)
+#
+
+
+def jacobian(
+    inputs: InputSchema,
+    jac_inputs: set[str],
+    jac_outputs: set[str],
+):
+    # convert all numbers and arrays to torch tensors
+    tensor_inputs = tree_map(map_lambda, inputs.model_dump())
+
+    # flatten the dict such that they can be accessed by paths and obtain the values
+    pos_inputs = flatten_with_paths(tensor_inputs, jac_inputs).values()
+
+    # create a positional function that accepts a list of values and returns a set of tuples
+    filtered_pos_eval = filter_pos_func(
+        evaluate, tensor_inputs, jac_inputs, jac_outputs, True
+    )
+
+    # calculate the jacobian
+    jacobian = torch.autograd.functional.jacobian(filtered_pos_eval, tuple(pos_inputs))
+
+    # rebuild the dictionary from the list of results
+    jax_dict = {}
+    for dy, dys in zip(jac_outputs, jacobian):
+        jax_dict[dy] = {}
+        for dx, dxs in zip(jac_inputs, dys):
+            jax_dict[dy][dx] = dxs
+
+    return jax_dict
+
+
+def jacobian_vector_product(
+    inputs: InputSchema,
+    jvp_inputs: set[str],
+    jvp_outputs: set[str],
+    tangent: dict[str, Any],
+):
+    # Make ordering of tangent identical to jvp_inputs
+    tangent = {key: tangent[key] for key in jvp_inputs}
+
+    # convert all numbers and arrays to torch tensors
+    tensor_inputs = tree_map(map_lambda, inputs.model_dump())
+    pos_tangent = tree_map(map_lambda, tangent).values()
+
+    # flatten the dictionaries such that they can be accessed by paths
+    pos_inputs = flatten_with_paths(tensor_inputs, jvp_inputs).values()
+
+    # create a positional function that accepts a list of values
+    filtered_pos_eval = filter_pos_func(
+        evaluate, tensor_inputs, jvp_inputs, jvp_outputs
+    )
+
+    return torch.func.jvp(filtered_pos_eval, tuple(pos_inputs), tuple(pos_tangent))[1]
+
+
+def vector_jacobian_product(
+    inputs: InputSchema,
+    vjp_inputs: set[str],
+    vjp_outputs: set[str],
+    cotangent_vector: dict[str, Any],
+):
+    # Make ordering of cotangent_vector identical to vjp_inputs
+    cotangent_vector = {key: cotangent_vector[key] for key in vjp_outputs}
+
+    # convert all numbers and arrays to torch tensors
+    tensor_inputs = tree_map(map_lambda, inputs.model_dump())
+    tensor_cotangent = tree_map(map_lambda, cotangent_vector)
+
+    # flatten the dictionaries such that they can be accessed by paths
+    pos_inputs = flatten_with_paths(tensor_inputs, vjp_inputs).values()
+
+    # create a positional function that accepts a list of values
+    filtered_pos_func = filter_pos_func(
+        evaluate, tensor_inputs, vjp_inputs, vjp_outputs
+    )
+
+    _, vjp_func = torch.func.vjp(filtered_pos_func, *pos_inputs)
+
+    vjp_vals = vjp_func(tensor_cotangent)
+    return dict(zip(vjp_inputs, vjp_vals))
+
+
+map_lambda = lambda x: torch.tensor(x) if isinstance(x, np.generic | np.ndarray) else x
```

Comment:
Let's give this a more telling name

```suggestion
to_tensor = lambda x: torch.tensor(x) if isinstance(x, np.generic | np.ndarray) else x
```